src/main/java/org/wso2/carbon/hash/provider/pbkdf2/PBKDF2HashProvider.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.hash.provider.pbkdf2;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.hash.provider.pbkdf2.constant.Constants;
+
+import org.wso2.carbon.user.core.exceptions.HashProviderException;
+import org.wso2.carbon.user.core.hash.HashProvider;
+
+import java.nio.charset.StandardCharsets;
+import java.security.NoSuchAlgorithmException;
+import java.security.spec.InvalidKeySpecException;
+import java.util.Base64;
+import java.util.Map;
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.PBEKeySpec;
+
+/**
+ * This class contains the implementation for the PBKDF2 hashing algorithm.
+ */
+public class PBKDF2HashProvider implements HashProvider {
+
+    private static final Log log = LogFactory.getLog(PBKDF2HashProvider.class);
+
+    @Override
+    public byte[] getHash(String value, String salt, Map<String, Object> metaProperties)
+            throws HashProviderException {
+
+        validateEmptyValue(value);
+        int iterationCount = getCount(metaProperties);
+        int dkLen = getDerivedKeyLength(metaProperties);
+        String pseudoRandomFunction = getPseudoRandomFunction(metaProperties);
+        return pbkdf2HashCalculation(value, salt, iterationCount, dkLen, pseudoRandomFunction);
+    }
+
+    @Override
+    public String getAlgorithm() {
+
+        return Constants.DEFAULT_HASHING_ALGORITHM;
+    }
+
+    /**
+     * This method is responsible for validating the value.
+     *
+     * @param value The value which needs to be hashed.
+     * @throws HashProviderException The custom exception which is thrown at validating the value.
+     */
+    private void validateEmptyValue(String value) throws HashProviderException {
+
+        if (StringUtils.isBlank(value)) {
+            throw new HashProviderException(
```

Comment:
We need to define a `HashProviderClientException` at the kernel level extending `HashProviderException` and throw it here since this is actually a client/user error.